### ISSUE TITLE
Neon should do any input, output shape check

### ISSUE THREAD

**Issue by mingwugmail** [NervanaSystems/neon#234]
https://groups.google.com/forum/?hl=en#!topic/neon-users/hWeVdMcm-O8

I'm a newbie try to play with Neon. I try to understand how the input & label are feed:

in examples/mnist_mlp.py

the output layer has 10 units:

layers.append(Affine(nout=10, init=init_norm, activation=Softmax()))

After load_mnist(), I print type(y_train), and y_train.shape, it's 

<type 'numpy.ndarray'> (60000,) of <type 'numpy.uint8'>

I just wonder how this single integer label is feed to the mapped to the 10 units? is there a hidden argmax() function somewhere? and it set desired_output[label] = 1.0?

Then I try to hack it: immediately after load_mnist(...), I add:

y_train = y_train.reshape((30000, 2))

And start to train the net, surprisingly it didn't complain anything, and run thru!

Epoch 0   [Train |████████████████████|  469/469  batches, 0.24 cost, 1.01s]
Epoch 1   [Train |████████████████████|  469/469  batches, 0.19 cost, 1.03s]
Epoch 2   [Train |████████████████████|  469/469  batches, 0.16 cost, 1.03s]
Epoch 3   [Train |████████████████████|  468/468  batches, 0.13 cost, 1.02s]
Epoch 4   [Train |████████████████████|  469/469  batches, 0.11 cost, 1.03s]
Epoch 5   [Train |████████████████████|  469/469  batches, 0.10 cost, 1.03s]
Epoch 6   [Train |████████████████████|  469/469  batches, 0.10 cost, 1.03s]
Epoch 7   [Train |████████████████████|  468/468  batches, 0.09 cost, 1.02s]
Epoch 8   [Train |████████████████████|  469/469  batches, 0.08 cost, 1.03s]
Epoch 9   [Train |████████████████████|  469/469  batches, 0.07 cost, 1.03s]
Misclassification error = 2.8%

This is amusing. I think Neon should report the len(X_train) and len(y_train) doesn't match, and I wonder how the train is done after my hack.

Next I added:

y_train += nclass

Again, surprisingly Neon didn't complain anything, and run thru!

Epoch 0   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.02s]
Epoch 1   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 2   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 3   [Train |████████████████████|  468/468  batches, 0.00 cost, 1.03s]
Epoch 4   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 5   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 6   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 7   [Train |████████████████████|  468/468  batches, 0.00 cost, 1.03s]
Epoch 8   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Epoch 9   [Train |████████████████████|  469/469  batches, 0.00 cost, 1.03s]
Misclassification error = 89.9%

I just wonder how the output label is mapped to the 10 units? even if it set desired_output[label] = 1.0, it should report index out of bound error.

Can anyone help me understand how the single integer label is feed to the mapped to the 10 units?

Thanks.

**Comment by mwoodson1** [comment]
Take a look at the **ArrayIterator** method if your curious but at a high level it converts your train labels to one-hot vectors on the fly. So the (6000,) vector you see are the actual labels(0-9) which get converted by the following call:

`train_set = ArrayIterator(X_train, y_train, nclass=nclass, lshape=(1, 28, 28))`

This will convert your training labels into vectors of size nclass with your input dimensions being (1,28,28). Hope that helps.

**Comment by mingwugmail** [comment]
OK, I saw there is a make_onehot param, but why there is no  index out of bound error after my hack:

y_train += nclass

Another experiment, I defined my own make_onehot func:

(X_train, y_train), (X_test, y_test), nclass = load_mnist(path=args.data_dir)

def make_onehot(y):
  one_hots = np.zeros((len(y), nclass))
  one_hots[np.arange(len(y)), y] = 1
  return one_hots

y_train = make_onehot(y_train)
y_test  = make_onehot(y_test )

train_set = DataIterator(X_train, y_train, make_onehot=False)
valid_set = DataIterator(X_test, y_test, make_onehot=False)

Is this equivalent to what Neon does internally?

When I run the code, it gives slightly bigger 
Misclassification error = 2.8%
Misclassification error = 2.9%

while the original code typically gives:
Misclassification error = 2.5%
Misclassification error = 2.6%

did my code miss something?

**Comment by mwoodson1** [comment]
To your first question adding nclass will not change the shape of y_train. All you did was shift the labels from 0-9 to 10-19. This is still essentially the same problem.

You're one hot code looks correct. Without looking at the internal code I can only assume the small difference in performance is due to how batching is handled between DataIterator and ArrayIterator. If you are unsatisfied you could take a look at those two methods to see what might be the cause but such a small performance difference shouldn't be of concern.

**Comment by mingwugmail** [comment]
I tried:

(X_train, y_train), (X_test, y_test), nclass = load_mnist(path=args.data_dir)
y_train[0] = nclass * 10
print y_train[0]

now it has 11 labels [0..9, 100], but it still run thru, didn't give me an index bound error.

**Comment by hanlin-nervana** [comment]
We can certainly add some input checking to the ArrayIterator. Thanks for your suggestion. For classification, ArrayIterator expects integer labels from (0, 1, ... K-1), where K is the number of classes. 

More information on what ArrayIterator expects and one-hot encoding can be found here: http://neon.nervanasys.com/docs/latest/loading_data.html.

**Comment by mingwugmail** [comment]
@mwoodson1 
""
You're one hot code looks correct.
""

I just checked the doc of ArrayIterator:

http://neon.nervanasys.com/docs/latest/generated/neon.data.dataiterator.ArrayIterator.html?highlight=arrayiterator

it says:

y (ndarray, shape:[# examples, 1], optional) – Labels corresponding to the input features. If absent, the input features themselves will be returned as target values (AutoEncoder)

My question is: does the y.shape ([# examples, 1]) is enforced? or does it matter?
Because after my code change of using make_onehot=False, my y.shape is ([# examples, 10]), or [# examples, 11]

Since Neon didn't check / complain anything, I have no idea if I did something wrong.
